src/validation.cpp

```diff
@@ -5083,19 +5083,22 @@ bool LoadMempool(CTxMemPool& pool)
         }
 
         // TODO: remove this try except in v0.22
+        std::map<uint256, uint256> unbroadcast_txids;
         try {
-          std::set<uint256> unbroadcast_txids;
           file >> unbroadcast_txids;
           unbroadcast = unbroadcast_txids.size();
-
-          for (const auto& txid : unbroadcast_txids) {
-            pool.AddUnbroadcastTx(txid);
-          }
         } catch (const std::exception&) {
           // mempool.dat files created prior to v0.21 will not have an
```

Comment:
looking into this- is there really a way that the CAutoFile `>>` operator could cause a silent/partial failure? as I mentioned previously the idea of this try catch is just for a smooth upgrade. To elaborate further: so we don't print `Failed to deserialize mempool data on disk: %s. Continuing anyway."` when everything actually went fine. the current intention is to remove this in 0.22, so if there's a possibility for `unbroadcast_txids` to have garbage data written, I'd like to think this through more carefully to find a long term solution.

Reply:
Yes I believe so? Fortunately for std::map we deserialize an element and then insert it into the map, but is a possibility that we say that there should be 10 elements and there are only 5 elements, or there are 5 elements and their are actually 10. In the event the file is corrupt, we should likely treat it as if the entire thing is corrupt, rather than continuing to process.

Perhaps we should add an exception to throw from Unserialize if there are not the correct amount of entries?



I'm not sure to the degree we need to worry about files on our own local being corrupt, and there are other forms of corruption that can occur. But because it is a behavior change, I'm noting it.

Reply:
to be clear the issue only comes up when you don't have enough elements or when you have a half element presently.